core/iwasm/libraries/lib-wasi-threads/test/spawn_stress_test.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2023 Amazon.com Inc. or its affiliates. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#ifndef __wasi__
+#error This example only compiles to WASM/WASI target
+#endif
+
+#include <assert.h>
+#include <errno.h>
+#include <math.h>
+#include <pthread.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+enum CONSTANTS {
+    NUM_ITER = 100000,
+    NUM_RETRY = 5,
+    MAX_NUM_THREADS = 8,
+};
+
+unsigned int prime_numbers = 0;
+
+bool
+is_prime(unsigned int num)
+{
+    for (unsigned int i = 2; i <= (unsigned int)(sqrt(num)); ++i) {
+        if (num % i == 0) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void *
+check_if_prime(void *value)
+{
+    unsigned int *num = (unsigned int *)(value);
+    usleep(1000);
+    if (is_prime(*num)) {
+        __atomic_fetch_add(&prime_numbers, 1, __ATOMIC_SEQ_CST);
+    }
+    return NULL;
+}
+
+unsigned int
+validate()
+{
+    unsigned int counter = 0;
+    for (unsigned int i = 2; i <= NUM_ITER; ++i) {
+        counter += is_prime(i);
+    }
+
+    return counter;
+}
+
+int
+main(int argc, char **argv)
+{
+    pthread_t threads[MAX_NUM_THREADS];
+    unsigned int args[MAX_NUM_THREADS];
+    double percentage = 0.1;
+
+    for (unsigned int factorised_number = 2; factorised_number < NUM_ITER;
+         ++factorised_number) {
+        if (factorised_number > NUM_ITER * percentage) {
+            fprintf(stderr, "Stress test is %d%% finished\n",
+                    (unsigned int)(percentage * 100));
+            percentage += 0.1;
+        }
+
+        unsigned int thread_num = factorised_number % MAX_NUM_THREADS;
+        if (threads[thread_num] != 0) {
+            assert(pthread_join(threads[thread_num], NULL) == 0);
+        }
+
+        args[thread_num] = factorised_number;
+        threads[thread_num] = 0;
+        usleep(1000);
+
+        int status_code = -1;
+        for (int tries = 0; status_code != 0 && tries < NUM_RETRY; ++tries) {
+            status_code = pthread_create(&threads[thread_num], NULL,
+                                         &check_if_prime, &args[thread_num]);
+            assert(status_code == 0 || status_code == EAGAIN);
+            if (status_code == EAGAIN) {
+                usleep(10000);
+            }
+        }
+
+        assert(status_code == 0 && "Thread creation should succeed");
+        assert(threads[thread_num] != 0);
+    }
+
+    for (int i = 0; i < MAX_NUM_THREADS; ++i) {
+        assert(pthread_join(threads[i], NULL) == 0);
```

Comment:
probably worth checking if `threads[i]` is not 0 before the `pthread_join()`

Reply:
they're guaranteed not to be if ITER_NUM is bigger than MAX_NUM_THREAD because of line 96, aren't they?

Reply:
But I'll add this check for visibility

Reply:
A simple example - if you set MAX_NUM_THREAD to 8 and ITER_NUM to 1, then most of the thread ids will be 0.